docs/tutorials/sparse/csr.md

```diff
@@ -0,0 +1,258 @@
+data = a.data
+# access indices array
+indices = a.indices
+# access indptr array
+indptr = a.indptr
+{'a.stype': a.stype, 'data':data, 'indices':indices, 'indptr':indptr}
+```
+
+## Storage Type Conversion
+
+* We can convert an NDArray to a CSRNDArray and vice versa by using the ``tostype`` function:
+
+```python
+# create a dense NDArray
+ones = mx.nd.ones((2,2))
+# cast the storage type from default to csr
+csr = ones.tostype('csr')
+# cast the storage type from csr to default
+dense = csr.tostype('default')
+{'csr':csr, 'dense':dense}
+```
+
+* We can also convert the storage type by using the ``cast_storage`` operator:
+
+```python
+# create a dense NDArray
+ones = mx.nd.ones((2,2))
+# cast the storage type to csr
+csr = mx.nd.sparse.cast_storage(ones, 'csr')
+# cast the storage type to default
+dense = mx.nd.sparse.cast_storage(csr, 'default')
+{'csr':csr, 'dense':dense}
+```
+
+## Copies
+
+* The `copy` method makes a deep copy of the array and its data, and `copyto` method or the slice
+operator `[]` to copy to an existing array.
+
+```python
+a = mx.nd.ones((2,2)).tostype('csr')
+b = a.copy()
+c = mx.nd.sparse.zeros('csr', (2,2))
+c[:] = a
+d = mx.nd.sparse.zeros('csr', (2,2))
+a.copyto(d)
+{'b is a': b is a, 'b.asnumpy()':b.asnumpy(), 'c.asnumpy()':c.asnumpy(), 'd.asnumpy()':d.asnumpy()}
+```
+
+* If the storage types of source array and destination array doesn't match,
+the storage type of destination array won't change when copying with `copyto` or
+the slice operator `[]`.
+
+```python
+e = mx.nd.sparse.zeros('csr', (2,2))
+f = mx.nd.sparse.zeros('csr', (2,2))
+g = mx.nd.ones(e.shape)
+e[:] = g
+g.copyto(f)
+{'e.stype':e.stype, 'f.stype':f.stype}
+```
+
+## Indexing and Slicing
+
+We can slice a CSRNDArray on axis 0 with operator `[]`, which copies the slices and returns a new CSRNDArray.
+
+```python
+a = mx.nd.array(np.arange(6).reshape(3,2)).tostype('csr')
+b = a[1:2]
+c = a[:].asnumpy()
+{'b':b, 'c':c}
+```
+
+## Sparse Operators and Storage Type Inference
+
+* Operators that have specialized implementation for sparse arrays can be accessed in ``mx.nd.sparse``.
+You can read the [mxnet.ndarray.sparse API documentation](mxnet.io/api/python/ndarray.html) to find
+what sparse operators are available.
+
+```python
+shape = (3, 4)
+data = [7, 8, 9]
+indptr = [0, 2, 2, 3]
+indices = [0, 2, 1]
+a = mx.nd.sparse.csr_matrix(data, indptr, indices, shape) # a csr matrix as lhs
+rhs = mx.nd.ones((4, 1))      # a dense vector as rhs
+out = mx.nd.sparse.dot(a, rhs)  # invoke sparse dot operator specialized for dot(csr, dense)
+{'out':out}
+```
+
+* For any sparse operator, the storage type of output array is inferred based on inputs. You can either read
+the documentation or inspect the `stype` attribute of output array to know what storage type is inferred:
+
+```python
+b = a * 2  # b will be a CSRNDArray since zero multiplied by 2 is still zero
+c = a + 1  # c will be a dense NDArray
+{'b.stype':b.stype, 'c.stype':c.stype}
```

Comment:
You say: b will be a CSRNDArray, but I see it as NDArray only. Am I interpreting things correctly?

>>> b = a * 2  # b will be a CSRNDArray since zero multiplied by 2 is still zero
>>> c = a + 1  # c will be a dense NDArray
>>> {'b.stype':b.stype, 'c.stype':c.stype}
{'c.stype': 'default', 'b.stype': 'default'}
>>> a.stype
'csr'
>>> b.stype
'default'  <======= NOT a CSRNDArray.
>>> c.stype
'default'
>>> b
[[ 14.   0.  16.   0.]
 [  0.   0.   0.   0.]
 [  0.  18.   0.   0.]]
<NDArray 3x4 @cpu(0)> <======= NOT a CSRNDArray.
>>> a
[[ 7.  0.  8.  0.]
 [ 0.  0.  0.  0.]
 [ 0.  9.  0.  0.]]
<CSRNDArray 3x4 @cpu(0)>

Reply:
Currently it results in a dense NDArray because Chris's PR is not merged in, as mentioned in summary